scd/token.c

```diff
@@ -35,8 +53,257 @@ struct scd_token_data {
+		goto connection_err;
+	} else {
+		ERROR("Unexpected event");
+		goto connection_err;
+	}
+
+	return;
+
+connection_err:
+	token->token_data->tctrl->events = list_remove(token->token_data->tctrl->events, io);
+	event_remove_io(io);
+	event_io_free(io);
+
+	token->token_data->tctrl->cfd = -1;
+	if (close(fd) < 0)
+		WARN_ERRNO("Failed to close connected control socket");
+	return;
+}
+
+/**
+ * Event callback for accepting incoming connections on the listening socket.
+ *
+ * @param fd	    file descriptor of the listening socket
+ *		    from which incoming connectionis should be accepted
+ * @param events    event flags
+ * @param io	    pointer to associated event_io_t struct
+ * @param data	    pointer to this scd_control_t struct
+  */
+static void
+scd_tokencontrol_cb_accept(int fd, unsigned events, UNUSED event_io_t *io, void *data)
+{
+	scd_token_t *token = (scd_token_t *)data;
+
+	if (events & EVENT_IO_READ) {
+		token->token_data->tctrl->cfd = accept(fd, NULL, 0);
+		if (-1 == token->token_data->tctrl->cfd) {
+			WARN("Could not accept tokencontrol connection");
+			return;
+		}
+		DEBUG("Accepted tokencontrol connection %d", token->token_data->tctrl->cfd);
+
+		fd_make_non_blocking(token->token_data->tctrl->cfd);
+
+		DEBUG("MADE tokenctrl_cfd non-blocking");
+
+		event_io_t *event = event_io_new(token->token_data->tctrl->cfd, EVENT_IO_READ,
+						 scd_tokencontrol_cb_recv_message, data);
+
+		DEBUG("2");
+
+		token->token_data->tctrl->events =
+			list_append(token->token_data->tctrl->events, event);
+
+		event_add_io(event);
+
+		DEBUG("3");
+
+	} else if (events & EVENT_IO_EXCEPT) {
+		TRACE("EVENT_IO_EXCEPT on socket %d, closing...", fd);
+	} else {
+		ERROR("Unexpected event");
+	}
+	return;
+}
+
+static int UNUSED
+scd_tokencontrol_new(scd_token_t *token)
+{
+	ASSERT(token && token->token_data);
+
+	TRACE("scd_tokencontrol_new");
+
+	token->token_data->tctrl = mem_new0(tctrl_t, 1);
+	IF_NULL_GOTO_ERROR(token->token_data->tctrl, err);
+
+	token->token_data->tctrl->lsock_path = mem_printf(
+		"%s/%s", SCD_TOKENCONTROL_SOCKET, uuid_string(token->token_data->token_uuid));
+	IF_NULL_GOTO_ERROR(token->token_data->tctrl->lsock_path, err);
+
+	token->token_data->tctrl->lsock = sock_unix_create_and_bind(
+		SOCK_STREAM | SOCK_NONBLOCK, token->token_data->tctrl->lsock_path);
+	if (token->token_data->tctrl->lsock < 0) {
+		WARN("Could not create and bind UNIX domain socket");
+		goto err;
+	}
+	if (listen(token->token_data->tctrl->lsock, SCD_TOKENCONTROL_SOCK_LISTEN_BACKLOG) < 0) {
+		WARN_ERRNO("Could not listen on new control sock");
+		close(token->token_data->tctrl->lsock);
+		goto err;
+	}
+
+	event_io_t *event = event_io_new(token->token_data->tctrl->lsock, EVENT_IO_READ,
+					 scd_tokencontrol_cb_accept, token);
+	token->token_data->tctrl->events = list_append(token->token_data->tctrl->events, event);
+	event_add_io(event);
+
+	return 0;
+
+err:
+	mem_free(token->token_data->tctrl);
```

Comment:
if(token->token_data->tctrl) mem_free()

Reply:
not necessasry, macro mem_free() already contains the check